proposals/module-linking/Binary.md

```diff
@@ -0,0 +1,219 @@
+            0x61 mt:moduletype                  ->        mt
+            0x62 it:instancetype                ->        it
+
+functype ::= rt_1:resulttype rt_2:resulttype    ->        rt_1 -> rt-2
+
+moduletype ::=
+  i*:vec(import) e*:vec(exporttype)             ->        {imports i, exports e}
+
+instancetype ::= e*:vec(exporttype)             ->        {exports e}
+
+exporttype ::= nm:name d:importdesc             ->        {name nm, desc d}
+```
+
+referencing:
+
+* [`resulttype`](https://webassembly.github.io/spec/core/binary/types.html#binary-resulttype)
+* [`import`](https://webassembly.github.io/spec/core/binary/modules.html#binary-import)
+
+**Validation**
+
+* each `importdesc` is valid according to import section
+* Types can only reference preceding type definitions. This forces everything to
+  be a DAG and forbids cyclic types.
+
+## Import Section updates
+
+Updates to
+[`importdesc`](https://webassembly.github.io/spec/core/binary/modules.html#binary-importdesc)
+
+```
+# note that `0x01 0xc0` in MVP-wasm specifies a 1-byte module field of the
+# string 0xc0, but the 0xc0 byte is not valid utf-8, so this was not a valid
+# MVP import
+import ::=
+    ...
+    mod:name  0x01 0xc0 d:importdesc            ->    {module mod, desc d}
+
+importdesc ::=
+    ...
+    0x05 x:typeidx                              ->    module x
+    0x06 x:typeidx                              ->    instance x
+```
+
+The `import` production is modified to allow optionally dropping the `name`
+field. This is done by encoding an invalid utf-8 string so this is an invalid
+MVP import.
+
+**Validation**
+
+* the `func x` production ensures the type `x` is indeed a function type
+* the `module x` production ensures the type `x` is indeed a module type
+* the `instance x` production ensures the type `x` is indeed an instance type
+
+## Module section (100)
+
+A new module section is added
+
+```
+modulesec ::=  t*:section_100(vec(typeidx))           ->        t*
+```
+
+**Validation**
+
+* Each type index `t` points to a module type (e.g. not a function type)
+* This defines the locally defined set of modules, adding to the module index
+  space.
+
+## Instance section (101)
+
+A new section defining local instances
+
+```
+instancesec ::=  i*:section_101(vec(instancedef))     ->    i*
+
+instancedef ::= m:moduleidx e*:vec(exportdesc)        ->    {instantiate m, imports e}
+```
+
+This defines instances in the module, appending to the instance index space.
+Each instance definition declares the module it's instantiating as well as the
+items used to instantiate that instance.
+
+**Validation**
+
+* The type index `m` must point to a module type.
+* The length of `e*` must be the same as the number of imports the module type
+  has
+* The type of each element of `e*` must be a subtype of the type that it's being
+  matched with. Matching happens pairwise with the list of imports on the module
+  type for `m`.
```

Comment:
Could you say something like "The real validations rules are being hashed out in [#7](https://github.com/WebAssembly/module-linking/issues/7) that allows the order-independent subtyping mentioned in the explainer.  To get started, simple, conservative order-dependent rules are used: ..."?

Reply:
To make sure I understand, the discussion on #7 is only applicable to subtyping, right? You'll always need to pass the exact number of imports to the `instantiate` instruction that the module type you're instantiating specifies, right? (linking to #7 sounds good at least in the subtyping section below)

Reply:
Well, at the end of the discussion, the idea is that there are also tweaks to how positional contexts like `instantiate` work.  In particular, `instantiate` would supply `number-of-unique-import-strings` operands, so that, for any import string, there was always an unambiguous value, and that value's type would be validated to be a subtype of each import with the same string.